test/fetchInterceptor.spec.js

```diff
@@ -0,0 +1,345 @@
+
+      fetch('http://localhost:5000/401/1?invalidate=true').then(response => {
+        expect(response.status).to.be.equal(200);
+        return response.json();
+      })
+        .then(data => {
+          expect(data.value).to.be.equal('1');
+          done();
+        })
+        .catch(error => {
+          done(error);
+        });
+    });
+
+    it('should fetch multiple simultaneous requests successfully with access token expired', function (done) {
+      // set expired access token
+      fetchInterceptor.authorize('refresh_token', 'token1');
+
+      Promise.all([
+        fetch('http://localhost:5000/401/1?duration=100'),
+        fetch('http://localhost:5000/401/2?duration=300'),
+        fetch('http://localhost:5000/401/3?duration=100'),
+      ])
+        .then(results => {
+          return {first: results[0], second: results[1], third: results[2]}
+        })
+        .then(responses => {
+          expect(responses.first.status).to.be.equal(200);
+          expect(responses.second.status).to.be.equal(200);
+          expect(responses.third.status).to.be.equal(200);
+
+          done();
+        })
+        .catch(error => {
+          done(error);
+        });
+    });
+
+    it('should fetch multiple requests successfully with access token expired', function (done) {
+      // set expired access token
+      fetchInterceptor.authorize('refresh_token', 'token1');
+
+      Promise.all([
+        fetch('http://localhost:5000/401/1?duration=100'),
+        Promise.resolve(delayPromise(100)).then(() => fetch('http://localhost:5000/401/2?duration=300')),
+        Promise.resolve(delayPromise(200)).then(() => fetch('http://localhost:5000/401/2?duration=100')),
+      ])
+        .then(results => {
+          return {first: results[0], second: results[1], third: results[2]}
+        })
+        .then(responses => {
+
+          expect(responses.first.status).to.be.equal(200);
+          expect(responses.second.status).to.be.equal(200);
+          expect(responses.third.status).to.be.equal(200);
+
+          done();
+        })
+        .catch(error => {
+          done(error);
+        });
+    });
+  });
+
+  describe('refresh token is invalid', () => {
+    beforeEach(done => {
+      fetchInterceptor.configure(configuration());
+      fetchInterceptor.authorize('invalid_refresh_token');
+
+      server.start(done);
+    });
+
+    afterEach(done => {
+      server.stop(done);
+    });
+
+    it('should propagate 401 for single request', function (done) {
+      fetch('http://localhost:5000/401/1').then(response => {
+        const tokens = fetchInterceptor.getAuthorization();
+
+        expect(response.status).to.be.equal(401);
+
+        expect(tokens.accessToken).to.be.null;
+        expect(tokens.refreshToken).to.be.null;
+
+        done();
+      })
+        .catch((error) => {
+          done(error);
+        });
+    });
+
+    it('should propagate 401 for multiple requests', function (done) {
```

Comment:
What happens here? Can we test what happens if one request gets multiple 401s?
1. Request is emitted with expired token
2. Response is 401
3. token is refreshed, request is reemitted
4. Response is 401
5. token is refreshed, request is reemitted
6. Response is 401
7. token is refreshed, request is reemitted
...